forgerock-authenticator/src/main/java/org/forgerock/android/auth/AuthenticatorManager.java

```diff
@@ -251,25 +255,78 @@ PushNotification getNotificationByMessageId(String messageId) {
         return storageClient.getNotificationByMessageId(messageId);
     }
 
-    void registerForRemoteNotifications(String newDeviceToken) throws AuthenticatorException {
-        if(this.deviceToken == null) {
-            this.deviceToken = newDeviceToken;
-            this.pushFactory = new PushFactory(context, storageClient, newDeviceToken);
-            this.notificationFactory = new NotificationFactory(storageClient);
-            PushResponder.getInstance(storageClient);
+    void registerForRemoteNotifications(String deviceToken) throws AuthenticatorException {
+        if(this.pushDeviceTokenManager.getDeviceTokenId() == null) {
+            this.pushDeviceTokenManager.setDeviceToken(deviceToken);
         } else {
-            if(this.deviceToken.equals(newDeviceToken)) {
+            if(this.pushDeviceTokenManager.getDeviceTokenId().equals(deviceToken)) {
                 Logger.warn(TAG, "The SDK was already initialized with this device token: %s",
-                        newDeviceToken);
+                        deviceToken);
                 throw new AuthenticatorException("The SDK was already initialized with the FCM device token.");
             } else {
                 Logger.warn(TAG, "The SDK was initialized with a different deviceToken: %s, however a new " +
-                        "device token (%s) was received.", this.deviceToken, newDeviceToken);
-                throw new AuthenticatorException("The SDK was initialized with a different deviceToken.");
+                        "device token (%s) was received.", this.pushDeviceTokenManager.getDeviceTokenId(), deviceToken);
+                throw new AuthenticatorException("The SDK was initialized with a different deviceToken. " +
+                        "Use FRAClient#updateDeviceToken method to update the device token.");
             }
         }
     }
 
+    void updateDeviceToken(String newDeviceToken, FRAListener<Boolean> listener) {
```

Comment:
I think it makes sense to provide this convenient method for developer, most of the time they want to update all the mechanisms, however we may want to improve the signature. If reusing FRAListener, we may want to use FRAListener<Void>, throw exception when one or more mechanism is failed. You may want to reuse AuthenticatorException.

It is kind of wierd that the developer implement onException() but it will never be called even failed.

Reply:
I agree, I have updated the code following your suggestion.